File: Environments/MicroRTS/MicroRTS-master/src/ai/mcts/uct/RQBonusMCTS.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ai.mcts.uct;

import ai.RandomScriptedAI;
import ai.abstraction.UpdateableAbstractAction;
import ai.abstraction.pathfinding.AStarPathFinding;
import ai.core.AI;
import java.util.List;
import java.util.ArrayList;
import rts.GameState;
import ai.evaluation.HPDifference;
import ai.evaluation.BasicOneZero;
import ai.portfolio.portfoliogreedysearch.Kiter;
import ai.portfolio.portfoliogreedysearch.NOKAV;
import ai.portfolio.portfoliogreedysearch.DependentUnitScript;
import ai.selection.UCB;
import ai.selection.SelectionFunction;
import ai.evaluation.EvaluationFunction;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.util.Arrays;
import java.util.Map;
import rts.units.UnitTypeTable;


public class RQBonusMCTS extends BasicMCTS {
    /**
    * A constant used in computing the sigmoid function of the bonuses.
    */
    private double _k = 1.0;

    /**
    * The average combined distance of the partial game tree traversal and the playout during
    * an iteration.
    */
    private double _averageDistance = 0;

    /**
    * The sample standard deviation of the sampled simulation lengths.
    */
    private double _sampleStandardRelativeDeviation = 0;

    /**
    * Like averageDistance, but all the lengths of playouts where player 0 lost get substituted with zeros.
    */
    private double _averageWinDistance = 0;

    /**
    * A list of all the simulation distances encountered so far.
    * Has to be kept for the sample standard deviation to be computable.
    */
    private List<Double> _distances = new ArrayList<>();

    /**
    * A list of all the final state qualities encountered so far.
    * Has to be kept for the sample standard deviation to be computable.
    */
    private List<Double> _qualities = new ArrayList<>();

    /**
    * A list of the results of all the playouts until now. Used when computing alpha.
    */
    private List<Double> _playoutResults = new ArrayList<>();

    /**
    * The average quality of final states encountered in playouts.
    */
    private double _averageQuality = 0;

    /**
    * The sample standard deviation of the sampled simulation lengths.
    */
    private double _sampleStandardQualitativeDeviation = 0;

    /**
    * Like averageQuality, but all the lengths of playouts where player 0 lost get substituted with zeros.
    */
    private double _averageWinQuality = 0;

    /**
    * Determines whether the relative bonus is supposed to be added to the result.
    */
    private boolean relativeBonusEnabled;

    /**
    * Determines whether the qualitative bonus is supposed to be added to the result.
    */
    private boolean qualitativeBonusEnabled;
    
    /**
     * An additional evaluation function used to compute state quality when
     * qualitative bonus is enabled.
     */
    private HPDifference hpdiff = new HPDifference();
    
    /**
     * Stores the depth of the last playout.
     */
    private int playoutDepth = 0;
    
    public RQBonusMCTS(UnitTypeTable utt) {
        super(100,-1,100,10,
             new RandomScriptedAI(Arrays.asList(new NOKAV(new AStarPathFinding()), new Kiter(new AStarPathFinding(), 3))),
             new BasicOneZero(),
             new UCB(),
             Arrays.asList(new NOKAV(new AStarPathFinding()), new Kiter(new AStarPathFinding(), 3)),
             "rq_bonus_mcts",
             "");
    }  
    
    public RQBonusMCTS(int available_time, int max_playouts, int lookahead, int max_depth, RandomScriptedAI policy, List<DependentUnitScript> scripts, double k, boolean rE, boolean qE, String name, String bmrkID) {
        super(available_time, max_playouts, lookahead, max_depth, policy, new BasicOneZero(), new UCB(), scripts, name, bmrkID);
        
        _k = k;
        relativeBonusEnabled = rE;
        qualitativeBonusEnabled = qE;
    }
    
    private RQBonusMCTS(
            int available_time, int max_playouts, int lookahead, int max_depth, RandomScriptedAI policy, List<DependentUnitScript> scripts,
            double k, boolean rE, boolean qE,
            double avgD, double sSRD, double avgWD, List<Double> dists, List<Double> quals, List<Double> pRes, double avgQ, double sSQD, double avgWQ,
            String name, String bmrkID,
            SelectionFunction sf, EvaluationFunction ef
    ) {
        super(available_time, max_playouts, lookahead, max_depth, policy, ef, sf, scripts, name, bmrkID);
        
        _k = k;
        relativeBonusEnabled = rE;
        qualitativeBonusEnabled = qE;
        
        _averageDistance = avgD;
        _sampleStandardRelativeDeviation = sSRD;
        _averageWinDistance = avgWD;
        
        _distances = new ArrayList<Double>(dists);
        _qualities = new ArrayList<Double>(quals);
        _playoutResults = new ArrayList<Double>(pRes);
        
        _averageQuality = avgQ;
        _sampleStandardQualitativeDeviation = sSQD;
        _averageWinQuality = avgWQ;
    }
    
    @Override
    public AI clone() {
        return new RQBonusMCTS(
                TIME_BUDGET, ITERATIONS_BUDGET, MAXSIMULATIONTIME, MAX_TREE_DEPTH, defaultPolicy, scripts,
                _k, relativeBonusEnabled, qualitativeBonusEnabled,
                _averageDistance, _sampleStandardRelativeDeviation, _averageWinDistance,
                _distances, _qualities, _playoutResults,
                _averageQuality, _sampleStandardQualitativeDeviation, _averageWinQuality,
                name, bmrkID,
                sf, ef
        );
    }
    
    // Overriden so that the appropriate values can be reset.
    @Override
    public void computeDuringOneGameFrame() throws Exception {
        // Reset variables between moves.
        _averageDistance = 0;
        _sampleStandardRelativeDeviation = 0;
        _averageWinDistance = 0;
        _distances = new ArrayList<>();
        _qualities = new ArrayList<>();
        _playoutResults = new ArrayList<>();
        _averageQuality = 0;
        _sampleStandardQualitativeDeviation = 0;
        _averageWinQuality = 0;
        
        long start = System.currentTimeMillis();
        nPlayouts = 0;
        long cutOffTime = start + TIME_BUDGET;
        if (TIME_BUDGET<=0) cutOffTime = 0;

        PrintStream ps = null;
        try {
            ps = new PrintStream(new FileOutputStream(path + name + "/" + bmrkID + "_" + battleName + "_" + randomBattleString + ".txt", true));
        }       
        catch (FileNotFoundException e) {
            throw new Error("Can't log time and depth.");
        }
        
        while(true) {                      
            long roundStart = System.nanoTime();
            if (cutOffTime>0 && System.currentTimeMillis() > cutOffTime) break;
            nPlayouts++;
            if (ITERATIONS_BUDGET>0 && nPlayouts>ITERATIONS_BUDGET) break;
            monteCarloRun(playerForThisComputation, cutOffTime);
            logTimeAndDepth(System.nanoTime() - roundStart, ps);
            
            // If there is only one possible action at the root, there is no reason
            // to continue running the algorithm.
            if (root.children.size() == 1 && !root.hasMoreActions) {
                break;
            }
        }
        
        total_cycles_executed++;
    }
    
    // Overridden so that the relative and qualitative bonuses can be computed.
    @Override
    public double monteCarloRun(int player, long cutOffTime) throws Exception {
        MCTSNode leaf = root.UCTSelectLeaf(player, 1-player, MAX_TREE_DEPTH, true);
        if (leaf.depth > maxDepth) {
            maxDepth = leaf.depth;
        }    

        if (leaf!=null) {
            GameState gs2 = leaf.gs.clone();
            simulate(gs2, gs2.getTime() + MAXSIMULATIONTIME, leaf.getAssignedAbstractActions());

            int time = gs2.getTime() - gs_to_start_from.getTime();
            double evaluation = ef.evaluate(player, 1-player, gs2);

            _playoutResults.add(evaluation);
            
            // Compute relative bonus, if it is enabled.
            if (relativeBonusEnabled)
            {
                int distance = leaf.depth + playoutDepth;
                
                if (_sampleStandardRelativeDeviation > 0)
                {
                    // Compute the relative bonus based on simulation length.
                    double lambdaR = Bonus(_averageDistance - (distance), _sampleStandardRelativeDeviation);
                    evaluation += evaluation * GetAlpha(_averageDistance, _averageWinDistance, _distances, _sampleStandardRelativeDeviation) * lambdaR;
                }

                // Update the list of encountered distances, the mean, the mean of winning distances and the sample standard deviation.
                _distances.add((double)distance);
                _averageDistance = (_averageDistance * (nPlayouts - 1) + distance) / nPlayouts;
                _averageWinDistance = (_averageWinDistance * (nPlayouts - 1) + Math.max(0, Math.signum(evaluation)) * distance) / nPlayouts;

                if (nPlayouts > 1)
                {
                    _sampleStandardRelativeDeviation = 0;
                    for (double d : _distances)
                    {
                        _sampleStandardRelativeDeviation += (d - _averageDistance) * (d - _averageDistance);
                    }
                    _sampleStandardRelativeDeviation = (float)Math.sqrt(_sampleStandardRelativeDeviation / (nPlayouts - 1));
                }
            }

            // Compute qualitative bonus, if it is enabled.
            if (qualitativeBonusEnabled)
            {
                double quality = hpdiff.evaluate(player, 1-player, gs2);

                if (_sampleStandardQualitativeDeviation > 0)
                {
                    // Compute the qualitative bonus based on simulation length.
                    double lambdaR = Bonus(quality - _averageQuality, _sampleStandardQualitativeDeviation);
                    evaluation += evaluation * GetAlpha(_averageQuality, _averageWinQuality, _qualities, _sampleStandardQualitativeDeviation) * lambdaR;
                }

                // Update the list of encountered qualities, the mean, the mean of winning qualities and the sample standard deviation.
                _qualities.add(quality);
                _averageQuality = (_averageQuality * (nPlayouts - 1) + quality) / nPlayouts;
                _averageWinQuality = (_averageQuality * (nPlayouts - 1) + Math.max(0, Math.signum(evaluation)) * quality) / nPlayouts;

                if (nPlayouts > 1)
                {
                    _sampleStandardQualitativeDeviation = 0;
                    for (double q : _qualities)
                    {
                        _sampleStandardQualitativeDeviation += (q - _averageQuality) * (q - _averageQuality);
                    }
                    _sampleStandardQualitativeDeviation = (float)Math.sqrt(_sampleStandardQualitativeDeviation / (nPlayouts - 1));
                }
            }

            Backpropagate(leaf, evaluation);
            
            total_runs++;
            total_runs_this_move++;
            return evaluation;
        } else {
            // no actions to choose from :)
            System.err.println(this.getClass().getSimpleName() + ": claims there are no more leafs to explore...");
            return 0;
        }
    }
    
    // Overriden in order to add playoutDepth incrementation.
    @Override
    public void simulate(GameState gs, int time, Map<Long, UpdateableAbstractAction> aaas) throws Exception {
        boolean gameover = false;
        playoutDepth = 0;
        RandomScriptedAI randomPlayer1 = (RandomScriptedAI) defaultPolicy.clone();
        randomPlayer1.setAbstractActions(aaas);
        RandomScriptedAI randomPlayer2 = (RandomScriptedAI) defaultPolicy.clone();
        randomPlayer2.setAbstractActions(aaas);

        do{
            if (gs.isComplete()) {
                gameover = gs.cycle();
            } else {
                gs.issue(defaultPolicy.getAction(0, gs));
                gs.issue(defaultPolicy.getAction(1, gs));
                playoutDepth += 2;
            }
        }while(!gameover && gs.getTime()<time);   
    }
    
    /**
     * Computes the bonus from the sample standard deviation and offset from the mean.
     */
    private double Bonus(double offsetFromMean, double sampleStandardDeviation)
    {
        double lambda = offsetFromMean / sampleStandardDeviation;
        return -1 + 2 / (1 + Math.exp(-_k * lambda));
    }

    /**
    * Computes the alpha multiplier used when computing the bonus.
    */
    private double GetAlpha(double averageMetric, double averageWinMetric, List<Double> metricList, double sampleStandardDeviation)
    {
        double covariance = 0;
        for (int i = 0; i < metricList.size(); ++i)
        {
            double winM = Math.max(0, _playoutResults.get(i)) * metricList.get(i);
            covariance += (winM - averageWinMetric) * (metricList.get(i) - averageMetric);
        }
        covariance /= nPlayouts - 1;

        return Math.abs(covariance / sampleStandardDeviation);
    }
    
    @Override
    public String toString()
    {
        return "RQ-Bonus MCTS, k: " + _k + ", relative bonus: " + relativeBonusEnabled + ", qualitative bonus: " + qualitativeBonusEnabled + ", playouts: " + ITERATIONS_BUDGET;
    }
}

File: Environments/COG AI Core/CMS/Micro/Search/MCTS/RQBonusMcts.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using CMS.ActionGenerators;
using CMS.Actions;
using CMS.GameStateEval;
using CMS.Players;
using CMS.Playout;
using System.Linq;
using System.Text;
using System.Xml;
using QuickGraph;
using QuickGraph.Serialization;

namespace CMS.Micro.Search.MCTS
{
    /// <summary>
    /// A variant that applies a bonus to the result of a playout based on its length and/or the quality of the result.
    /// </summary>
    public class RQBonusMcts : Mcts
    {
        /// <summary>
        /// A constant used in computing the sigmoid function of the bonuses.
        /// </summary>
        private readonly double _k = 1.0;

        /// <summary>
        /// The simulation distance from the root to the node created in the expansion phase.
        /// </summary>
        private int _traversalDistance = 0;

        /// <summary>
        /// The average combined distance of the partial game tree traversal and the playout during
        /// an iteration.
        /// </summary>
        private double _averageDistance = 0;

        /// <summary>
        /// The sample standard deviation of the sampled simulation lengths.
        /// </summary>
        private double _sampleStandardRelativeDeviation = 0;

        /// <summary>
        /// Like averageDistance, but all the lengths of playouts where player 0 lost get substituted with zeros.
        /// </summary>
        private double _averageWinDistance = 0;

        /// <summary>
        /// A list of all the simulation distances encountered so far.
        /// Has to be kept for the sample standard deviation to be computable.
        /// </summary>
        private List<double> _distances = new List<double>();

        /// <summary>
        /// A list of all the final state qualities encountered so far.
        /// Has to be kept for the sample standard deviation to be computable.
        /// </summary>
        private List<double> _qualities = new List<double>();

        /// <summary>
        /// A list of the results of all the playouts until now. Used when computing alpha.
        /// </summary>
        private List<double> _playoutResults = new List<double>();

        /// <summary>
        /// The average quality of final states encountered in playouts.
        /// </summary>
        private double _averageQuality = 0;

        /// <summary>
        /// The sample standard deviation of the sampled simulation lengths.
        /// </summary>
        private double _sampleStandardQualitativeDeviation = 0;

        /// <summary>
        /// Like averageQuality, but all the lengths of playouts where player 0 lost get substituted with zeros.
        /// </summary>
        private double _averageWinQuality = 0;

        /// <summary>
        /// Determines whether the relative bonus is supposed to be added to the result.
        /// </summary>
        private bool relativeBonusEnabled;

        /// <summary>
        /// Determines whether the qualitative bonus is supposed to be added to the result.
        /// </summary>
        private bool qualitativeBonusEnabled;

        public RQBonusMcts(
            IActionGenerator actionGenerator,
            int maxPlayouts,
            List<Player> players,
            double k,
            string bmrkID,
            int playoutRoundLimit = int.MaxValue,
            bool relativeBonusEnabled = false,
            bool qualitativeBonusEnabled = false) : base(actionGenerator, maxPlayouts, players, bmrkID, playoutRoundLimit, relativeBonusEnabled ? qualitativeBonusEnabled ? "rq_bonus_mcts" : "r_bonus_mcts" : "q_bonus_mcts")
        {
            _k = k;
            this.relativeBonusEnabled = relativeBonusEnabled;
            this.qualitativeBonusEnabled = qualitativeBonusEnabled;
        }

        // Overriden so that the appropriate values are reset between moves.
        internal override ICollection<ActionCms> GetActions(GameEnvironment environment)
        {
            // Reset variables between moves.
            _averageDistance = 0;
            _sampleStandardRelativeDeviation = 0;
            _averageWinDistance = 0;
            _distances = new List<double>();
            _qualities = new List<double>();
            _playoutResults = new List<double>();
            _averageQuality = 0;
            _sampleStandardQualitativeDeviation = 0;
            _averageWinQuality = 0;

            TotalIterCount = 0;
            maxDepth = 0;
            var rootActionState = new ActionStatePair { Environment = environment };
            IEnumerable<ActionStatePair> rootChildActions = _actionGenerator.EnumerateActions(rootActionState.Environment);

            var root = new MctsNode(null, rootActionState, rootChildActions, 0);

            using (var sw = new StreamWriter(path + name + "/" + bmrkID + "_" + battleName + "_" + randomBattleString + ".txt", true))
            {
                while (TotalIterCount < _maxPlayouts)
                {
                    long timeStamp = Stopwatch.GetTimestamp();

                    ++TotalIterCount;
                    // Selection and Expansion
                    MctsNode selectedNode = SelectAndExpand(root);

                    // Playout
                    GameEnvironment selectedStateClone = selectedNode.CurrentActionState.Environment.DeepCloneState();
                    GameResult result = Game.Playout(selectedStateClone, _players, _playoutRoundLimit);
                    double value = EvaluateGame(result, selectedStateClone);

                    // Backpropagation
                    BackpropagateResults(selectedNode, value);

                    if (selectedNode.depth > maxDepth)
                    {
                        maxDepth = selectedNode.depth;
                    }

                    LogTimeAndDepth(1_000_000_000.0 * (Stopwatch.GetTimestamp() - timeStamp) / Stopwatch.Frequency, sw);
                }
            }

            MctsNode bestNode = SelectBestChild(root, environment.GameState.ActivePlayer, (n, p, b) => p == 0 ? n.Value / n.VisitedCount : -n.Value / n.VisitedCount);
            var actions = bestNode.CurrentActionState.PlayerAction;

            return actions;
        }

        // Overridden so that the depth of the tree can be measured using the _traversalDistance variable.
        protected override MctsNode SelectAndExpand(MctsNode node)
        {
            Debug.Assert(node != null);
            _traversalDistance = 0;

            while (true)
            {
                // If node is not fully expanded - generate new child node with this new action and return it
                MctsNode newChild;
                if (node.TryGetNextChild(_actionGenerator, out newChild))
                {
                    return newChild;
                }
                else if (node.IsTerminal)
                {
                    return node;
                }

                // Find child node with highest value and call Selection recursively
                int player = node.CurrentActionState.Environment.GameState.ActivePlayer;
                MctsNode bestChild = SelectBestChild(node, player, Ucb);

                if (bestChild == null)
                    return null;

                node = bestChild;
                _traversalDistance++;
            }
        }

        // Adds a bonus to the default 1, 0, -1 evaluation.
        protected override double EvaluateGame(GameResult result, GameEnvironment environment)
        {
            double reward;

            if (result.Winner == GameState.WINNER_NONE || result.IsDraw)
            {
                reward = 0.0;
            }
            else if (result.Winner == 0)
            {
                reward = 1.0;
            }
            else
            {
                // We lost or are in bad position
                reward = -1.0;
            }

            _playoutResults.Add(reward);

            if (relativeBonusEnabled)
            {
                int distance = _traversalDistance + result.RoundCount;

                if (_sampleStandardRelativeDeviation > 0)
                {
                    // Compute the relative bonus based on simulation length.
                    double lambdaR = Bonus(_averageDistance - (distance), _sampleStandardRelativeDeviation);
                    double alpha = GetAlpha(_averageDistance, _averageWinDistance, _distances, _sampleStandardRelativeDeviation);
                    reward += reward * alpha * lambdaR;
                }

                // Update the list of encountered distances, the mean, the mean of winning distances and the sample standard deviation.
                _distances.Add(distance);
                _averageDistance = (_averageDistance * (TotalIterCount - 1) + distance) / TotalIterCount;
                _averageWinDistance = (_averageWinDistance * (TotalIterCount - 1) + Math.Max(0, Math.Sign(reward)) * distance) / TotalIterCount;

                if (TotalIterCount > 1)
                {
                    _sampleStandardRelativeDeviation = 0;
                    foreach (int d in _distances)
                    {
                        _sampleStandardRelativeDeviation += (d - _averageDistance) * (d - _averageDistance);
                    }
                    _sampleStandardRelativeDeviation = (float)Math.Sqrt(_sampleStandardRelativeDeviation / (TotalIterCount - 1));
                }
            }

            // Compute qualitative bonus, if it is enabled.
            if (qualitativeBonusEnabled)
            {
                double quality = environment.GameState.Units[0].Values.Sum(x => x.Hull)
                           - environment.GameState.Units[1].Values.Sum(x => x.Hull);

                if (_sampleStandardQualitativeDeviation > 0)
                {
                    // Compute the qualitative bonus based on simulation length.
                    double lambdaR = Bonus(quality - _averageQuality, _sampleStandardQualitativeDeviation);
                    reward += reward * GetAlpha(_averageQuality, _averageWinQuality, _qualities, _sampleStandardQualitativeDeviation) * lambdaR;
                }

                // Update the list of encountered qualities, the mean, the mean of winning qualities and the sample standard deviation.
                _qualities.Add(quality);
                _averageQuality = (_averageQuality * (TotalIterCount - 1) + quality) / TotalIterCount;
                _averageWinQuality = (_averageQuality * (TotalIterCount - 1) + Math.Max(0, Math.Sign(reward)) * quality) / TotalIterCount;

                if (TotalIterCount > 1)
                {
                    _sampleStandardQualitativeDeviation = 0;
                    foreach (float q in _qualities)
                    {
                        _sampleStandardQualitativeDeviation += (q - _averageQuality) * (q - _averageQuality);
                    }
                    _sampleStandardQualitativeDeviation = (float)Math.Sqrt(_sampleStandardQualitativeDeviation / (TotalIterCount - 1));
                }
            }

            return reward;
        }

        public override string ToString()
        {
            return $"RQ-Bonus MCTS, k: {_k}, relative bonus: {relativeBonusEnabled}, qualitative bonus: {qualitativeBonusEnabled}, playouts: {_maxPlayouts}";
        }

        /// <summary>
        /// Computes the bonus from the sample standard deviation and offset from the mean.
        /// </summary>
        private double Bonus(double offsetFromMean, double sampleStandardDeviation)
        {
            double lambda = offsetFromMean / sampleStandardDeviation;
            return -1 + 2 / (1 + Math.Exp(-_k * lambda));
        }

        /// <summary>
        /// Computes the alpha multiplier used when computing the bonus.
        /// </summary>
        private double GetAlpha(double averageMetric, double averageWinMetric, List<double> metricList, double sampleStandardDeviation)
        {
            double covariance = 0;
            for (int i = 0; i < metricList.Count; ++i)
            {
                double winM = Math.Max(0, _playoutResults[i]) * metricList[i];
                covariance += (winM - averageWinMetric) * (metricList[i] - averageMetric);
            }
            covariance /= TotalIterCount - 1;

            return Math.Abs(covariance / sampleStandardDeviation);
        }
    }
}
